### ISSUE TITLE
File size 0 with resumable.js pause and resume

### ISSUE THREAD

**Issue by riaan53** [vsivsi/meteor-file-collection#28]
Hi,

When I pause the upload with Media.resumable.pause() and resume it again with Media.resumable.upload() the upload finishes with no errors on client or server but the file size is zero. Any ideas why this is happening?

Regards,
Riaan

**Comment by vsivsi** [comment]
@riaan53 Hi, I'm just recently back from vacation. Thanks for this question and for the other work you've done on other topics. I'll start working on these tomorrow. Thanks for your patience!

**Comment by riaan53** [comment]
Hi,

Just some extra information.

When working with the resumable file objects directly you can pause and resume individual files (its not in resumable.js documentation). For example: Media.resumable.files[0].pause(true) and Media.resumable.files[0].pause(false). It is necessary to call Media.resumable.upload() as well for the current paused file to start again which is fine. Im very glad about this feature and it works perfectly with file-collection!

When using abort and resume on the resumable file objects I also get a zero file size without any errors like in my first comment.

So its just the pause/resume on the whole queue and the abort/retry on resumable.js file objects that results in a zero file size.

**Comment by vsivsi** [comment]
Hi, The way file-collection works, a zero-size file is always created first to provide a "target" for the HTTP upload method. So somewhere in your code, you are doing the equivalent of this line of code:
https://github.com/vsivsi/meteor-file-sample-app/blob/master/sample.coffee#L35

The file remains zero-size until the upload finishes. 

In the case of a basic HTTP POST/PUT, that occurs just after the input stream from the HTTP request body ends.

When using resumable.js, something different happens because the upload is broken into a series of HTTP POST requests which can occur in any order (and can be paused, etc).  So in this case, the zero length file is created and locked (using gridfs-locks).  Then, as the chunks of the file are uploaded, each chunk is written to the gridFS store as a temporary file with just that single chunk. Once all of the chunks for a file have been received, then the data chunks are properly ordered in the grid collection and assigned to the zero-length file. That causes it's file size to update and a new md5 sum to be calculated by the underlying gridFS driver.

So, now with all of that out of the way, the only way that what you are seeing *should* happen is if for some reason, the resumable.js upload is not actually uploading every chunk of the file under the conditions you describe (pausing all uploads).  This can easily be verified, because all of the one chunk temporary files for that upload should be sitting in the file collection.  You can find them on the server as:
`fileCollection.find({ 'metadata._Resumable': { $exists: true })`.

The `metadata._Resumable` attribute contains an object with fields like:
```
metadata._Resumable.resumableIdentifier
metadata._Resumable.resumableChunkNumber
metadata._Resumable.resumableTotalChunks
metadata._Resumable.resumableTotalSize
metadata._Resumable.resumableFilename
```
My guess is that if you go look in the database at the chunks for one of these failed uploads, you'll find that some chunks are still missing, and so the file hasn't been transformed from zero-length because the upload hasn't really completed.  If this turns out to be the case, then the bug lies in resumable.js.  If all of the chunks are present and accounted for, then the bug is in file-collection.

**Comment by vsivsi** [comment]
@riaan53  What's the status of this issue on your end?  Resolved?

**Comment by riaan53** [comment]
Hi @vsivsi  ,

Sorry I havnt got back to this issue yet. 

Just one question about something else - does file-collection support chunk overrides for resumable.js? 

From resumable.js documentation: 

*You should allow for the same chunk to be uploaded more than once; this isn’t standard behaviour, but on an unstable network environment it could happen, and this case is exactly what Resumable.js is designed for.*

**Comment by vsivsi** [comment]
Hmm. A duplicate chunk will be written. Question is if it will overwrite the previous chunk, or land in a new one.  My guess is that it will land in a new chunk with the same number, and that may foul things up as currently implemented. I've never observed this to actually happen because we use the resumable.js GET before POST functionality to check for the presence of a chunk on the server before sending, but there could be a race or some other network problem that causes a chunk to arrive twice, I suppose. 

This line is where the new chunk is created, and it doesn't check to see if a matching chunk already exists. https://github.com/vsivsi/meteor-file-collection/blob/master/resumable_server.coffee#L155

A `find` could be added for that filename before the 'upsertStream` And if an existing chunk file is found, its `_id` added to the `upsertStream` call to trigger an overwrite rather than a new file of the same name (with a different _id). 

Should be an easy fix.

**Comment by vsivsi** [comment]
@riaan53  I just granted you push access to the file-collection repo.  Please develop on branches and consult with me before merging into the master branch and to coordinate on releases to Atmosphere.

**Comment by riaan53** [comment]
Thank you @vsivsi ! :)

**Comment by vsivsi** [comment]
@riaan53  At long last, this will be fixed in version 1.1.0.  the `dev_1.1` branch has the changes, due for release very soon.

**Comment by vsivsi** [comment]
Fixed in 1.1.0
